### ISSUE TITLE
Free variables

### ISSUE THREAD

**Issue by josd** [w3c/N3#19]
Should we support free variables (i.e., that are not quantified)?

In the resolution of https://github.com/w3c/N3/issues/7 a variable name is a Well-Known IRI.
When a variable name is not declared via @forAll or @forSome it is a free variable (also called a parameter).

Here is a concrete example:

`PREFIX var: <http://example.com/.well-known/genid-n3/okUV7D18aoxPqAoy9zlZmgAyE4Q#>`
`:s :p var:y.`

**Comment by william-vw** [comment]
Could you please give an example within a use case (what would the semantics of its substitution be)?

**Comment by josd** [comment]
A typical use case is to "free" the quantified variables in N3 formulae e.g. test.n3

```
PREFIX : <http://example.org/try#>

{?S :p ?O} => {?O :q _:X}.
```

is transformed (e.g. using eye --nope --no-qvars test.n3 --pass-all) into

```
PREFIX var: <http://josd.github.io/.well-known/genid-n3/jB_hXbhapVRVWB5X6Rtdj3gRExU#>
PREFIX : <http://example.org/try#>

{var:sk_0 :p var:sk_1} => {var:sk_1 :q var:sk_2}.
```

This result can be further processed by other reasoning runs.
A reasoner can also provide a mechanism to go back to quantified variables.
There are no substitutions for free variables i.e. there is no way to bind them.

**Comment by doerthe** [comment]
I think that additionally to this use case free variables are always useful when people normally abuse blank nodes. If you take for example the case of reification, the problem in representing
```
:Wikipedia :says {:Melville :wrote :MobyDick}.
```
by using rdf reification:
```
:Wikipedia :says _:x.
_:x a rdf:Statement; 
      rdf:subject :Melville; 
     rdf:predicate :wrote; 
     rdf:object :MobyDick.
```
(yes, the example is taken from the other thread)

If we use this construction, we have the problem that the `_:x` is not just a variable, it is an existentially quantified variable. We do not say that wikipedia states a certain triple, we say that there exists a triple wikipedia states and that this triple has certain subeject, predicate and object. 

By introducing free variables we can at least get rid of the implicit quantifier.

**Comment by dckc** [comment]
You're skolemizing universals (`?S`) as well as existentials? (`_:X`). I'm not sure how to make sense of that.

**Comment by josd** [comment]
"Freeing" the quantified variables in N3 formulae makes them logically totally
different formulae i.e. they are no longer "sentences" (see https://en.wikibooks.org/wiki/Formal_Logic/Predicate_Logic/Free_and_Bound_Variables)
That said, I just checked all our current use cases and there is no use case
anymore that requires eye --no-qvars so I can live without free variables.

**Comment by doerthe** [comment]
I agree that it does not make sense to skolemize universals. What does make sense is to have a separate namespace for quantified variables in my opinion, only to not mix up variables and constants. 

Additionally we can discuss about free variables in general (not skolemized, just free).

**Comment by william-vw** [comment]
@josd do you think this issue can be closed, or perhaps deferred until later?

(just cleaning up a bit .. we have a lot of issues still open but I think some of them have been resolved)

**Comment by josd** [comment]
yes @william-vw this issue can be deferred until later
